execute/window.go

```diff
@@ -127,3 +131,74 @@ func truncateByMonths(t Time, d Duration) Time {
 	ts = time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
 	return values.ConvertTime(ts)
 }
+
+func WindowStart(timestamp Time, every, offset Duration) Time {
+	// separate month value from timestamp
+	_, m, _ := timestamp.Time().Date()
+	totalMonths := int64(m)
+
+	// calculate months for window start
+	modMonths := modulo(totalMonths, every.Months())
+	off := modulo(offset.Months(), every.Months())
+	months := calculateStart(totalMonths, every.Months(), modMonths, off)
+
+	// subtract the total months from the timestamp
+	nano := int64(timestamp) - (months * monthInNSecs)
+
+	// and calculate nanoseconds for window start
+	modNano := modulo(nano, every.Nanoseconds())
+	offsetNano := modulo(offset.Nanoseconds(), every.Nanoseconds())
+	nsecs := calculateStart(nano, every.Nanoseconds(), modNano, offsetNano)
+
+	// add the start months back to the start nanoseconds
+	return values.Time(nsecs + (months * monthInNSecs))
```

Comment:
There may be an easier way to get the right types here. Definitely open to changing this if so.